Docs: Backend edits (#2059)
grafana/plugin-tools: docusaurus/docs/key-concepts/backend-plugins/index.md

sidebar_position: 1.5
---

# Plugin backend system

Introduced in Grafana v7.0, plugins with backend components allow you to integrate Grafana with virtually anything and offer custom visualizations. The plugin backend system is based on HashiCorp's [Go Plugin System over RPC](https://github.com/hashicorp/go-plugin) and supports server-side plugin elements. The Grafana server launches each plugin backend as a subprocess and communicates with it over [gRPC](https://grpc.io/).

:::note

> A plugin with a backend usually requires a frontend component as well. For example, some backend data source plugins need query editor components on the frontend.

Reviewer: This was not accurate in the current documentation, a frontend component is always necessary.

```suggestion
A plugin with a backend requires a frontend component as well. For example, all data source plugins need a query editor component on the frontend.
```

:::

> ## Benefits for plugin development

Reviewer: ```suggestion
## Benefits for plugin backend development
```

To add custom metrics to instrument your plugin's backend, refer to [Implement metrics in your plugin](../../how-to-guides/data-source-plugins/add-logs-metrics-traces-for-backend-plugins.md#implement-metrics-in-your-plugin).

### Streaming

The streaming capability allows a plugin's backend to handle data source queries that are streaming. For more information, refer to the tutorial for a [streaming data source plugin](../../tutorials/build-a-streaming-data-source-plugin.md).

## Data communication model

Grafana uses a communication model where you can opt in to instance management to simplify the development process. If you do, then all necessary information (configuration) is provided in each request to a plugin backend, allowing the plugin to fulfill the request and return a response. This model simplifies for plugin authors not having to keep track of or request additional state to fulfill a request.

## Caching and connection pooling

> Grafana provides instance management in the plugin's backend SDK to ease working with multiple configured Grafana data sources or apps, referred to as instances. This allows a plugin to simply keep state cleanly separated between instances. The SDK makes sure to optimize plugin resources by caching said instances in memory until their configuration changes in Grafana. Refer to the [Data source plugin backend tutorial](/tutorials/build-a-data-source-backend-plugin) or the [App with backend documentation](/how-to-guides/app-plugins/add-backend-component), which shows how to use the instance management for data source and app plugins.

Reviewer: ```suggestion
Grafana provides instance management as part of the plugin SDK to ease working with multiple configured Grafana data sources or apps, referred to as instances. This allows a plugin to simply keep state cleanly separated between instances. The SDK makes sure to optimize plugin resources by caching said instances in memory until their configuration changes in Grafana. Refer to the [Data source plugin backend tutorial](/tutorials/build-a-data-source-backend-plugin) or the [App with backend documentation](/how-to-guides/app-plugins/add-backend-component), which shows how to use the instance management for data source and app plugins.
```
